# ebi-metagenomics/assembly-analysis-pipeline: Citations

## MGnify

> Richardson L, Allen B, Baldi G, Beracochea M, Bileschi ML, Burdett T, et al. MGnify: the microbiome sequence data analysis resource in 2023 [Internet]. Vol. 51, Nucleic Acids Research. Oxford University Press (OUP); 2022. p. D753–9. Available from: http://dx.doi.org/10.1093/nar/gkac1080

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [Diamond](http://dx.doi.org/10.1038/s41592-021-01101-x)

  > Buchfink B, Reuter K, Drost HG. Sensitive protein alignments at tree-of-life scale using DIAMOND. Vol. 18, Nature Methods. Springer Science and Business Media LLC; 202p. 366–8. Available from: http://dx.doi.org/10.1038/s41592-021-01101-x

- [Infernal](http://dx.doi.org/10.1093/bioinformatics/btt509)

  > Nawrocki EP, Eddy SR. Infernal 1.1: 100-fold faster RNA homology searches. Vol. 29, Bioinformatics. Oxford University Press (OUP); 2013. p. 2933–5. Available from: http://dx.doi.org/10.1093/bioinformatics/btt509

- [Combined Gene Caller](http://dx.doi.org/10.1093/nar/gkac1080)

  > Richardson L, Allen B, Baldi G, Beracochea M, Bileschi ML, Burdett T, et al. MGnify: the microbiome sequence data analysis resource in 2023. Vol. 51, Nucleic Acids Research. Oxford University Press (OUP); 2022. p. D753–9. Available from: http://dx.doi.org/10.1093/nar/gkac1080

- [Krona](http://dx.doi.org/10.1186/1471-2105-12-385)

  > Ondov BD, Bergman NH, Phillippy AM. Interactive metagenomic visualization in a Web browser. Vol. 12, BMC Bioinformatics. Springer Science and Business Media LLC; 201Available from: http://dx.doi.org/10.1186/1471-2105-12-385

- [Pyrodigal](http://dx.doi.org/10.21105/joss.04296)

  > Larralde M. Pyrodigal: Python bindings and interface to Prodigal, an efficient method for gene prediction in prokaryotes. Vol. 7, Journal of Open Source Software. The Open Journal; 2022. p. 4296. Available from: http://dx.doi.org/10.21105/joss.04296

- [dbCAN](http://dx.doi.org/10.1093/nar/gkad328)

  > Zheng J, Ge Q, Yan Y, Zhang X, Huang L, Yin Y. dbCAN3: automated carbohydrate-active enzyme and substrate annotation. Vol. 51, Nucleic Acids Research. Oxford University Press (OUP); 2023. p. W115–2Available from: http://dx.doi.org/10.1093/nar/gkad328

- [InterProScan](http://dx.doi.org/10.1093/bioinformatics/btu031)

  > Jones P, Binns D, Chang HY, Fraser M, Li W, McAnulla C, et al. InterProScan 5: genome-scale protein function classification. Vol. 30, Bioinformatics. Oxford University Press (OUP); 2014. p. 1236–40. Available from: http://dx.doi.org/10.1093/bioinformatics/btu031

- [catpack](http://dx.doi.org/10.1186/s13059-019-1817-x)

  > von Meijenfeldt FAB, Arkhipova K, Cambuy DD, Coutinho FH, Dutilh BE. Robust taxonomic classification of uncharted microbial sequences and bins with CAT and BAT. Vol. 20, Genome Biology. Springer Science and Business Media LLC; 2019. Available from: http://dx.doi.org/10.1186/s13059-019-1817-x

- [FragGeneScanRs](http://dx.doi.org/10.1186/s12859-022-04736-5)

  > Van der Jeugt F, Dawyndt P, Mesuere B. FragGeneScanRs: faster gene prediction for short reads. Vol. 23, BMC Bioinformatics. Springer Science and Business Media LLC; 2022. Available from: http://dx.doi.org/10.1186/s12859-022-04736-5

- [SanntiS](http://dx.doi.org/10.1101/2023.05.23.540769)

  > Sanchez S, Rogers JD, Rogers AB, Nassar M, McEntyre J, Welch M, et al. Expansion of novel biosynthetic gene clusters from diverse environments using SanntiS. Cold Spring Harbor Laboratory; 2023. Available from: http://dx.doi.org/10.1101/2023.05.23.540769

- [DRAM](http://dx.doi.org/10.1093/nar/gkaa621)

  > Shaffer M, Borton MA, McGivern BB, Zayed AA, La Rosa SL, Solden LM, et al. DRAM for distilling microbial metabolism to automate the curation of microbiome function. Vol. 48, Nucleic Acids Research. Oxford University Press (OUP); 2020. p. 8883–900. Available from: http://dx.doi.org/10.1093/nar/gkaa621

- [Genome properties](http://dx.doi.org/10.1093/nar/gky1013)

  > Richardson LJ, Rawlings ND, Salazar GA, Almeida A, Haft DR, Ducq G, et al. Genome properties in 2019: a new companion database to InterPro for the inference of complete functional attributes. Vol. 47, Nucleic Acids Research. Oxford University Press (OUP); 2018. p. D564–72. Available from: http://dx.doi.org/10.1093/nar/gky1013

- [tabix](http://dx.doi.org/10.1093/bioinformatics/btp352)

  > Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, et al. The Sequence Alignment/Map format and SAMtools. Vol. 25, Bioinformatics. Oxford University Press (OUP); 2009. p. 2078–9. Available from: http://dx.doi.org/10.1093/bioinformatics/btp352

- [antiSMASH](http://dx.doi.org/10.1093/nar/gkab335)

  > Blin K, Shaw S, Kloosterman AM, Charlop-Powers Z, van Wezel GP, Medema MH, et al. antiSMASH 6.0: improving cluster detection and comparison capabilities. Vol. 49, Nucleic Acids Research. Oxford University Press (OUP); 202p. W29–35. Available from: http://dx.doi.org/10.1093/nar/gkab335

- [HMMER](http://dx.doi.org/10.1371/journal.pcbi.1002195)

  > Eddy SR. Accelerated Profile HMM Searches. Pearson WR, editor. Vol. 7, PLoS Computational Biology. Public Library of Science (PLoS); 201p. e1002195. Available from: http://dx.doi.org/10.1371/journal.pcbi.1002195

- [QUAST](http://dx.doi.org/10.1093/bioinformatics/btt086)

  > Gurevich A, Saveliev V, Vyahhi N, Tesler G. QUAST: quality assessment tool for genome assemblies. Vol. 29, Bioinformatics. Oxford University Press (OUP); 2013. p. 1072–5. Available from: http://dx.doi.org/10.1093/bioinformatics/btt086

- [SeqKit](http://dx.doi.org/10.1371/journal.pone.0163962)

  > Shen W, Le S, Li Y, Hu F. SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation. Zou Q, editor. Vol. 11, PLOS ONE. Public Library of Science (PLoS); 2016. p. e0163962. Available from: http://dx.doi.org/10.1371/journal.pone.0163962

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
