Changes in module 'ebi-metagenomics/combinedgenecaller/merge'
--- modules/ebi-metagenomics/combinedgenecaller/merge/main.nf
+++ modules/ebi-metagenomics/combinedgenecaller/merge/main.nf
@@ -3,8 +3,8 @@
     label 'process_single'
 
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:1.0.1--pyhdfd78af_0':
-        'biocontainers/mgnify-pipelines-toolkit:1.0.1--pyhdfd78af_0' }"
+        "https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:${params.mpt_version}":
+        "biocontainers/mgnify-pipelines-toolkit:${params.mpt_version}" }"
 
     input:
     tuple val(meta), path(pyrodigal_gff, stageAs: "pyrodigal/"), path(pyrodigal_ffn, stageAs: "pyrodigal/"), path(pyrodigal_faa, stageAs: "pyrodigal/"), path(fgs_gff, stageAs: "fgsrs/"), path(fgs_ffn, stageAs: "fgsrs/"), path(fgs_faa, stageAs: "fgsrs/"), path(mask)

************************************************************
