Changes in module 'ebi-metagenomics/convertcmscantocmsearch'
--- modules/ebi-metagenomics/convertcmscantocmsearch/main.nf
+++ modules/ebi-metagenomics/convertcmscantocmsearch/main.nf
@@ -4,7 +4,10 @@
     label 'process_single'
 
     conda "${moduleDir}/environment.yml"
-    container "microbiome-informatics/mgnify-pipelines-toolkit:1.0.2"
+
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        "https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:${params.mpt_version}":
+        "biocontainers/mgnify-pipelines-toolkit:${params.mpt_version}" }"
 
     input:
     tuple val(meta), path(cmscan_tblout)

************************************************************
